### ISSUE TITLE
Error: file_put_contents(…): failed to open stream: No such file or directory at /var/www/html/lib/private/Files/Storage/Local.php#277

### ISSUE THREAD

**Issue by oncletom** [nextcloud/integration_google#19]
Hello,

I ran a Google Drive import today and encountered this error after a few minutes:

```
[PHP] Error: Error: file_put_contents(/var/www/html/data/admin/files/Importation de Google Drive/CartoBio/Entretiens utilisateurs/2020/04/17 CR Entretien Certis): failed to open stream: No such file or directory at /var/www/html/lib/private/Files/Storage/Local.php#277 at <<closure>>

 0. <<closure>>
    OC\Log\ErrorHandler::onError(2, "file_put_conten ... y", "/var/www/html/l ... p", 277, {path: "files/Im ... l})
 1. /var/www/html/lib/private/Files/Storage/Local.php line 277
    file_put_contents("/var/www/html/d ... s", null)
 2. /var/www/html/lib/private/Files/Storage/Wrapper/Wrapper.php line 257
    OC\Files\Storage\Local->file_put_contents("files/Importati ... s", null)
 3. /var/www/html/lib/private/Files/View.php line 1159
    OC\Files\Storage\Wrapper\Wrapper->file_put_contents("files/Importati ... s", null)
 4. /var/www/html/lib/private/Files/View.php line 696
    OC\Files\View->basicOperation("file_put_contents", "/admin/files/Im ... s", ["create","write"], null)
 5. /var/www/html/lib/private/Files/Node/Folder.php line 189
    OC\Files\View->file_put_contents("/admin/files/Im ... s", null)
 6. /var/www/html/custom_apps/integration_google/lib/Service/GoogleDriveAPIService.php line 291
    OC\Files\Node\Folder->newFile("2020/04/17 CR Entretien Certis", null)
 7. /var/www/html/custom_apps/integration_google/lib/Service/GoogleDriveAPIService.php line 212
    OCA\Google\Service\GoogleDriveAPIService->getFile("ya29.A0AfH6SMBX ... U", "admin", {kind: "drive#fi ... e}, {1d0KsPmWYAZlMym ... }}, OC\Files\Node\Folder {})
 8. /var/www/html/custom_apps/integration_google/lib/Service/GoogleDriveAPIService.php line 126
    OCA\Google\Service\GoogleDriveAPIService->importFiles("ya29.A0AfH6SMBX ... U", "admin", "Importation de Google Drive", 500000000, 0)
 9. /var/www/html/custom_apps/integration_google/lib/BackgroundJob/ImportDriveJob.php line 36
    OCA\Google\Service\GoogleDriveAPIService->importDriveJob("admin")
10. /var/www/html/lib/public/BackgroundJob/Job.php line 80
    OCA\Google\BackgroundJob\ImportDriveJob->run({user_id: "admin"})
11. /var/www/html/lib/public/BackgroundJob/QueuedJob.php line 48
    OCP\BackgroundJob\Job->execute(OC\BackgroundJob\JobList {}, OC\Log {})
12. /var/www/html/cron.php line 127
    OCP\BackgroundJob\QueuedJob->execute(OC\BackgroundJob\JobList {}, OC\Log {})

at 2020-10-30T10:16:13+00:00
```

The aforementioned file is a Google Doc file (text document).

The error happens next to this one:

```
[integration_google] Warning: Google API error : Client error: `GET https://www.googleapis.com/drive/v3/files/1atCOnjyUQzPCvv4oom9VlK6t7hwoX5mB_Z8hTm0BfAc?alt=media` resulted in a `403 Forbidden` response:
{
 "error": {
  "errors": [
   {
    "domain": "global",
    "reason": "fileNotDownloadable",
    "message": "Only files (truncated...)


at 2020-10-30T10:16:12+00:00
```

And the next error is the following:

```
[core] Error: OCP\Files\NotPermittedException: Could not create path at <<closure>>

0. /var/www/html/custom_apps/integration_google/lib/Service/GoogleDriveAPIService.php line 291
   OC\Files\Node\Folder->newFile("2020/04/17 CR Entretien Certis", null)
1. /var/www/html/custom_apps/integration_google/lib/Service/GoogleDriveAPIService.php line 212
   OCA\Google\Service\GoogleDriveAPIService->getFile("ya29.A0AfH6SMBX ... U", "admin", {kind: "drive#fi ... e}, {1d0KsPmWYAZlMym ... }}, OC\Files\Node\Folder {})
2. /var/www/html/custom_apps/integration_google/lib/Service/GoogleDriveAPIService.php line 126
   OCA\Google\Service\GoogleDriveAPIService->importFiles("ya29.A0AfH6SMBX ... U", "admin", "Importation de Google Drive", 500000000, 0)
3. /var/www/html/custom_apps/integration_google/lib/BackgroundJob/ImportDriveJob.php line 36
   OCA\Google\Service\GoogleDriveAPIService->importDriveJob("admin")
4. /var/www/html/lib/public/BackgroundJob/Job.php line 80
   OCA\Google\BackgroundJob\ImportDriveJob->run({user_id: "admin"})
5. /var/www/html/lib/public/BackgroundJob/QueuedJob.php line 48
   OCP\BackgroundJob\Job->execute(OC\BackgroundJob\JobList {}, OC\Log {})
6. /var/www/html/cron.php line 127
   OCP\BackgroundJob\QueuedJob->execute(OC\BackgroundJob\JobList {}, OC\Log {})

at 2020-10-30T10:16:13+00:00
```

I run Nextcloud 10.0.2 in a Docker container (`nextcloud:10-apache`) with SQLite backend.
The folder is a Google Drive shared folder (someone else shared it with me), but I have _Editor_ rights (read, write and delete files).

Since then, Nextcloud import reports 50 files imported (0%). The figures have not changed since a couple hours.

Also, when I connected my Google account, I accepted only the Drive permissions, as I have no intent to import contacts/calendar and photos — actually, my initial intent is to import/download only one folder and its whole content, not my entire Drive files (mix of personal, private and professional files).

**Comment by eneiluj** [comment]
Hey! Thanks a lot for the bug report.

So there are 2 issues here:
1. Google scandalously allows slashes in file/folder names -> Fixed
2. Google refuses that we download Google doc files. Anyway it's not stopping the import. Those files are just ignored and a warning pops (as you've seen).

v0.0.11 is out there. I successfully tested importing files and folders with a slash in their names. To push people to avoid putting slashes in file names, they are replaced with `-slash-` :grin:.

Could you try v0.0.11?

**Comment by oncletom** [comment]
Duly noted. I updated the extension, stopped the ongoing import and started it again.
I will check tomorrow the end result of this operation.

---

For the record, Google Documents can be "exported" via the [Files: export API](https://developers.google.com/drive/api/v3/reference/files/export). I used `application/vnd.openxmlformats-officedocument.wordprocessingml.document` as a MIME-Type to export a Google Doc into a `.docx` document. [Here is the exhaustive references of MIME types to export documents](https://developers.google.com/drive/api/v3/ref-export-formats).
